### PR TITLE
[Backport] Two phase segmented heap dump

### PR DESCRIPTION
**Pull request by y1yang0**
Summary: The minor-pause solution, which is presented in this PR, is a two phase segmented heap dump

1. Phase One(STW): Concurrent threads directly write data to multiple heap files.
2. Phase Two(Non-STW): Merge multiple heap files into one complete heap dump file.

Test Plan: manual

Reviewed-by : denghui.ddh, lei.yul, sanhong.lsh

Issue: https://github.com/dragonwell-project/dragonwell11/issues/630

### CODE REVIEW

File: test/hotspot/jtreg/serviceability/dcmd/gc/HeapDumpParallelTest.java [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+public class HeapDumpParallelTest {
+
+    private static void checkAndVerify(OutputAnalyzer dcmdOut, LingeredApp app, File heapDumpFile, boolean expectSerial) throws IOException {
+        dcmdOut.shouldHaveExitValue(0);
+        dcmdOut.shouldContain("Heap dump file created");
+        OutputAnalyzer appOut = new OutputAnalyzer(app.getProcessStdout());
+        appOut.shouldContain("[heapdump]");
+        String opts = Arrays.asList(Utils.getTestJavaOpts()).toString();
+        if (opts.contains("-XX:+UseSerialGC") || opts.contains("-XX:+UseEpsilonGC")) {
+            System.out.println("UseSerialGC detected.");
```
Comment (D-D-H) [inline test/hotspot/jtreg/serviceability/dcmd/gc/HeapDumpParallelTest.java:58]: doesn't match the condition

File: src/hotspot/share/services/heapDumper.cpp [hunk 2]
```diff
@@ -1479,6 +1565,138 @@ void HeapObjectDumper::do_object(oop o) {
...
+   ~DumperController() { delete _lock; }
+
+   void dumper_complete(DumpWriter* local_writer, DumpWriter* global_writer) {
+     MonitorLockerEx ml(_lock, Mutex::_no_safepoint_check_flag);
+     _complete_number++;
+     // propagate local error to global if any
+     if (local_writer->has_error()) {
+       global_writer->set_error(local_writer->error());
+     }
+     ml.notify();
```
Comment (D-D-H) [inline src/hotspot/share/services/heapDumper.cpp:1590]: notify is not necessary when  _complete_number != _dumper_number

Comment (y1yang0) [reply src/hotspot/share/services/heapDumper.cpp:1590]: Good catch, will ship with optimization patch for usptream.

**Review by D-D-H — APPROVED** [review]

File: src/hotspot/share/services/heapDumper.hpp [hunk 3]
```diff
@@ -90,6 +79,11 @@ class HeapDumper : public StackObj {
 
   static void dump_heap_from_oome()    NOT_SERVICES_RETURN;
 
+  // Parallel thread number for heap dump, initialize based on active processor count.
+  static uint default_num_of_dump_threads() {
```
Comment (yuleil) [inline src/hotspot/share/services/heapDumper.hpp:83]: Heapdump is executed at safepoint, why don't we use all CPU cores?

Comment (y1yang0) [reply src/hotspot/share/services/heapDumper.hpp:83]: This is derived from https://github.com/openjdk/jdk/pull/2261/files#diff-5373ebf2d7b4e9c8fcbb188dd0d7193d4af7e3e021de7e3e4dd61c9bb5803edfR249

Experiments show that using all CPUs does not achieve the best performance.

**Review by yuleil — APPROVED** [review]

**Pull request merged**